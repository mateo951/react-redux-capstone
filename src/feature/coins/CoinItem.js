/* eslint-disable camelcase */
import React from 'react';
import PropTypes from 'prop-types';
import style from './coinItem.module.css';

const CoinItem = ({
  coin_id, coin_name, coin_value, filter,
}) => (
  <>
    <li>
      <div className={style.coinItem}>
        <img alt="currency" src="https://img.icons8.com/fluency/35/000000/stack-of-coins.png" />
        <p>
          {`${coin_id} `}
          -
          {` ${coin_name}`}
        </p>
        <p>{filter === 'value' ? `$${coin_value}` : `${coin_value}%`}</p>
      </div>
    </li>
  </>
);
// const dispatch = useDispatch();
export default CoinItem;

CoinItem.propTypes = {
  coin_id: PropTypes.string.isRequired,
  coin_name: PropTypes.string.isRequired,
  coin_value: PropTypes.string.isRequired,
  filter: PropTypes.string.isRequired,
};
