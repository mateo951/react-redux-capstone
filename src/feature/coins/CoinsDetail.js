import { useSelector } from 'react-redux';
import { NavLink } from 'react-router-dom';
import style from './coins.module.css';

const CoinsDetail = () => {
  const { coins } = useSelector((state) => state.coins);
  const { active } = useSelector((state) => state.coins);
  return (
    <div className={style.detailContainer}>
      <NavLink className={style.backBttn} to="/">
        <img alt="Back button" src="https://img.icons8.com/ios-glyphs/30/ffffff/circled-left-2.png" />
      </NavLink>
      <p>Cryptocurrency</p>
      <div className={style.infoContainer}>
        <p className={style.detailInfo}>
          {`${coins[active].coin_id} `}
          -
          {` ${coins[active].coin_name}`}
        </p>
      </div>
      <p>Value</p>
      <div className={style.infoContainer}>
        <p className={style.detailInfo}>{`$${coins[active].coin_value}`}</p>
      </div>
      <p>Daily Fluctuation</p>
      <div className={style.infoContainer}>
        <p className={style.detailInfo}>{`${coins[active].coin_change}%`}</p>
      </div>
    </div>
  );
};
export default CoinsDetail;
