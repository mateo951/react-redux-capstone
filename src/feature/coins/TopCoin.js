import { useSelector } from 'react-redux';
import style from './topCoin.module.css';

const TopCoin = () => {
  const { coins } = useSelector((state) => state.coins);

  return (
    <>
      <div className={style.container}>
        <h3>
          Most Valuable -
          {' '}
          { coins !== undefined ? coins[0].coin_name : <p>Loading Crypto</p> }
        </h3>
        <div className={style.content}>
          <img alt="BTC" src="https://img.icons8.com/color/70/000000/bitcoin--v3.png" />
          { coins !== undefined ? `$${coins[0].coin_value}` : <p>Loading Value</p> }
        </div>
      </div>
    </>
  );
};

export default TopCoin;
