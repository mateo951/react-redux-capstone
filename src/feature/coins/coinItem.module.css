.coinItem {
  padding: 20px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.outerContainer {
  width: 50%;
}
