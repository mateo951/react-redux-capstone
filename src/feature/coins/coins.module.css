.coinsContainer {
  width: 100%;
  height: 500px;
}

.filtersContainer {
  display: flex;
  justify-content: space-between;
  align-items: center;
  font-size: 15px;
  color: white;
  background-color: #35548B;
  padding: 5px 10px;
}

.coinBox {
  display: flex;
  flex-wrap: wrap;
  color: white;
}

.route {
  width: 50%;
}

ul {
  text-decoration: none;
  list-style-type: none;
  margin: 0 !important;
  padding: 0 !important;
}

.coinsContainer ul a:nth-child(4n+1) {
  background-color: #4066AD;
  color: white;
}

.coinsContainer ul a:nth-child(4n) {
  background-color: #4066AD;
  color: white;
}

.coinsContainer ul a:nth-child(4n-1) {
  background-color: #4872C2;
  color: white;
}

.coinsContainer ul a:nth-child(4n-2) {
  background-color: #4872C2;
  color: white;
}

.detailContainer {
  background-color: #5689E5;
  height: 85vh;
  color: white;
  text-align: center;
  font-size: 25px;
  display: flex;
  flex-direction: column;
  align-items: center;
  row-gap: 10px;
  padding-top: 40px;
}

.infoContainer {
  width: 70%;
  height: 50px;
  background-color: #3E64AB;
  display: flex;
  justify-content: center;
  align-items: center;
}

.detailInfo {
  font-size: 17px;
}