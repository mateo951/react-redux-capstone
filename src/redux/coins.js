import axios from 'axios';

const apiURL = 'https://api.coincap.io/v2/assets?limit=10';
const GET_COINS = 'react-redux-capstone/coins/GET_COINS';
const UPDATE_FILTER = 'react-redux-capstone/coins/UPDATE_FILTER';
const SET_DETAIL = 'react-redux-capstone/coins/SET_DETAIL';
const HANDLE_FETCH_ERROR = 'react-redux-capstone/HANDLE_FETCH_ERROR';

const getCoins = (coins) => ({
  type: GET_COINS,
  payload: coins,
});

const handleFetchError = () => ({
  type: HANDLE_FETCH_ERROR,
});

export const setDetail = (index) => ({
  type: SET_DETAIL,
  payload: index,
});

export const fetchGetCoins = () => async (dispatch) => {
  try {
    const res = await axios.get(apiURL);
    const { data } = res.data;
    const coins = [];
    Object.entries(data).forEach(([key, value]) => {
      const coin = {
        coin_index: key,
        coin_rank: value.rank,
        coin_id: value.name,
        coin_name: value.symbol,
        coin_value: value.priceUsd.substring(0, 8),
        coin_change: value.changePercent24Hr.substring(0, 8),
      };
      coins.push(coin);
    });
    dispatch(getCoins(coins));
  } catch {
    dispatch(handleFetchError());
  }
};

const updateFilter = (activeFilter, coins) => ({
  type: UPDATE_FILTER,
  payload: { activeFilter, coins },
});

export const filterCoins = (filter, coins) => {
  if (filter !== 'value') {
    coins.sort((a, b) => b.coin_change - a.coin_change);
  } else {
    coins.sort((a, b) => b.coin_value - a.coin_value);
  }
  return updateFilter(filter, coins);
};

const reducer = (state = [], { type, payload }) => {
  switch (type) {
    case GET_COINS: {
      const activeFilter = 'value';
      return { ...state, coins: payload.slice(), activeFilter };
    }
    case UPDATE_FILTER: {
      const { coins } = payload;
      const { activeFilter } = payload;
      return { ...state, coins: coins.slice(), activeFilter };
    }
    case SET_DETAIL:
      return { ...state, active: payload };
    case HANDLE_FETCH_ERROR:
      return state;
    default:
      return state;
  }
};

export default reducer;
