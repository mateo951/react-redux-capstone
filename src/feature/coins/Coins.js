import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NavLink } from 'react-router-dom';
import { fetchGetCoins, filterCoins, setDetail } from '../../redux/coins';
import CoinItem from './CoinItem';
import style from './coins.module.css';

const Coins = () => {
  const dispatch = useDispatch();
  const { coins } = useSelector((state) => state.coins);
  useEffect(() => {
    if (coins === undefined || coins.length === 0) dispatch(fetchGetCoins());
  }, []);
  const { activeFilter } = useSelector((state) => state.coins);
  const handleChange = (event) => {
    dispatch(filterCoins(event.target.value, coins));
  };

  const handleDetail = (coinIndex) => {
    dispatch(setDetail(coinIndex));
  };

  return (
    <div className={style.coinsContainer}>
      <div className={style.filtersContainer}>
        <p>Top 10 Most Valuable Crypto</p>
        <select value={activeFilter} onChange={handleChange}>
          <option value="value">value</option>
          <option value="fluctuation">fluctuation</option>
        </select>
      </div>
      <ul className={style.coinBox}>
        {
          coins !== undefined ? coins.map((coin, index) => (
            <NavLink className={style.route} key={coin.coin_id} to="/coin-detail" onClick={() => { handleDetail(index); }}>
              <CoinItem
                key={coin.coin_id}
                coinId={coin.coin_id}
                coinName={coin.coin_name}
                coinValue={activeFilter === 'value' ? coin.coin_value : coin.coin_change}
                filter={activeFilter}
              />
            </NavLink>
          )) : <p>Loading Data</p>
        }
      </ul>
    </div>
  );
};

export default Coins;
