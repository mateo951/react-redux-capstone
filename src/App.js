import React from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import Navigation from './feature/navigation/Navigation';
import Coins from './feature/coins/Coins';
import './App.css';

function App() {
  return (
    <Router>
      <Navigation />
      <Switch>
        <Route exact path="/">
          <Coins />
        </Route>
      </Switch>
    </Router>
  );
}

export default App;
