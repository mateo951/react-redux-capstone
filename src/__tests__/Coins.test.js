import renderer from 'react-test-renderer';
import {
  render, screen,
} from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import Coins from '../feature/coins/Coins';
import App from '../App';
import store from '../redux/configureStore';
import { fetchGetCoins } from '../redux/coins';

describe('pages test snapshots', () => {
  it('testing detail page', async () => {
    await store.dispatch(fetchGetCoins());
    const coinsPage = renderer
      .create(
        <Provider store={store}>
          <App>
            <Coins />
          </App>
        </Provider>,
      )
      .toJSON();
    expect(coinsPage).toMatchSnapshot();
  });
});

describe('Rendered Data', () => {
  test('Displaying 10 crypto currencies', () => {
    render(
      <Provider store={store}>
        <Router>
          <App>
            <Coins />
          </App>
        </Router>
      </Provider>,
    );
    screen.queryAllByRole('link').forEach((role) => expect(role).toBeInTheDocument());
    expect(screen.queryAllByRole('link').length).toBe(11);
    // +1 do to the homepage link
  });
});
