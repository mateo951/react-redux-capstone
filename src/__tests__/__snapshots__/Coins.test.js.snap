// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages test snapshots testing detail page 1`] = `
Array [
  <header>
    <img
      alt="Mobile menu button"
      src="https://img.icons8.com/material-rounded/24/ffffff/menu--v1.png"
    />
    <a
      aria-current="page"
      className="active"
      href="/"
      onClick={[Function]}
      style={Object {}}
    >
      <h1
        className="title"
      >
        Cripto Live
      </h1>
    </a>
    <div
      className="icons"
    >
      <img
        alt="config button"
        src="https://img.icons8.com/external-those-icons-fill-those-icons/24/ffffff/external-gear-settings-those-icons-fill-those-icons-1.png"
      />
    </div>
  </header>,
  <div
    className="container"
  >
    <h3>
      Most Valuable -
       
      BTC
    </h3>
    <div
      className="content"
    >
      <img
        alt="BTC"
        src="https://img.icons8.com/color/70/000000/bitcoin--v3.png"
      />
      $61339.90
    </div>
  </div>,
  <div
    className="coinsContainer"
  >
    <div
      className="filtersContainer"
    >
      <p>
        Top 10 Most Valuable Crypto
      </p>
      <select
        onChange={[Function]}
        value="value"
      >
        <option
          value="value"
        >
          value
        </option>
        <option
          value="fluctuation"
        >
          fluctuation
        </option>
      </select>
    </div>
    <ul
      className="coinBox"
    >
      <a
        aria-current={null}
        className="route"
        href="/coin-detail"
        onClick={[Function]}
      >
        <li>
          <div
            className="coinItem"
          >
            <img
              alt="currency"
              src="https://img.icons8.com/fluency/35/000000/stack-of-coins.png"
            />
            <p>
              Bitcoin 
              -
               BTC
            </p>
            <p>
              $61339.90
            </p>
          </div>
        </li>
      </a>
      <a
        aria-current={null}
        className="route"
        href="/coin-detail"
        onClick={[Function]}
      >
        <li>
          <div
            className="coinItem"
          >
            <img
              alt="currency"
              src="https://img.icons8.com/fluency/35/000000/stack-of-coins.png"
            />
            <p>
              Ethereum 
              -
               ETH
            </p>
            <p>
              $4113.626
            </p>
          </div>
        </li>
      </a>
      <a
        aria-current={null}
        className="route"
        href="/coin-detail"
        onClick={[Function]}
      >
        <li>
          <div
            className="coinItem"
          >
            <img
              alt="currency"
              src="https://img.icons8.com/fluency/35/000000/stack-of-coins.png"
            />
            <p>
              Binance Coin 
              -
               BNB
            </p>
            <p>
              $482.4436
            </p>
          </div>
        </li>
      </a>
      <a
        aria-current={null}
        className="route"
        href="/coin-detail"
        onClick={[Function]}
      >
        <li>
          <div
            className="coinItem"
          >
            <img
              alt="currency"
              src="https://img.icons8.com/fluency/35/000000/stack-of-coins.png"
            />
            <p>
              Cardano 
              -
               ADA
            </p>
            <p>
              $2.154495
            </p>
          </div>
        </li>
      </a>
      <a
        aria-current={null}
        className="route"
        href="/coin-detail"
        onClick={[Function]}
      >
        <li>
          <div
            className="coinItem"
          >
            <img
              alt="currency"
              src="https://img.icons8.com/fluency/35/000000/stack-of-coins.png"
            />
            <p>
              Tether 
              -
               USDT
            </p>
            <p>
              $1.000209
            </p>
          </div>
        </li>
      </a>
      <a
        aria-current={null}
        className="route"
        href="/coin-detail"
        onClick={[Function]}
      >
        <li>
          <div
            className="coinItem"
          >
            <img
              alt="currency"
              src="https://img.icons8.com/fluency/35/000000/stack-of-coins.png"
            />
            <p>
              Solana 
              -
               SOL
            </p>
            <p>
              $196.1547
            </p>
          </div>
        </li>
      </a>
      <a
        aria-current={null}
        className="route"
        href="/coin-detail"
        onClick={[Function]}
      >
        <li>
          <div
            className="coinItem"
          >
            <img
              alt="currency"
              src="https://img.icons8.com/fluency/35/000000/stack-of-coins.png"
            />
            <p>
              XRP 
              -
               XRP
            </p>
            <p>
              $1.089344
            </p>
          </div>
        </li>
      </a>
      <a
        aria-current={null}
        className="route"
        href="/coin-detail"
        onClick={[Function]}
      >
        <li>
          <div
            className="coinItem"
          >
            <img
              alt="currency"
              src="https://img.icons8.com/fluency/35/000000/stack-of-coins.png"
            />
            <p>
              Polkadot 
              -
               DOT
            </p>
            <p>
              $43.44539
            </p>
          </div>
        </li>
      </a>
      <a
        aria-current={null}
        className="route"
        href="/coin-detail"
        onClick={[Function]}
      >
        <li>
          <div
            className="coinItem"
          >
            <img
              alt="currency"
              src="https://img.icons8.com/fluency/35/000000/stack-of-coins.png"
            />
            <p>
              Dogecoin 
              -
               DOGE
            </p>
            <p>
              $0.251038
            </p>
          </div>
        </li>
      </a>
      <a
        aria-current={null}
        className="route"
        href="/coin-detail"
        onClick={[Function]}
      >
        <li>
          <div
            className="coinItem"
          >
            <img
              alt="currency"
              src="https://img.icons8.com/fluency/35/000000/stack-of-coins.png"
            />
            <p>
              USD Coin 
              -
               USDC
            </p>
            <p>
              $1.000574
            </p>
          </div>
        </li>
      </a>
    </ul>
  </div>,
]
`;
